homeassistant/components/jlrincontrol/__init__.py

```diff
@@ -0,0 +1,193 @@
+    })
+}, extra=vol.ALLOW_EXTRA)
+
+
+def setup(hass, config):
+    """Set up the jlrpy component."""
+    import jlrpy
+
+    username = config[DOMAIN].get(CONF_USERNAME)
+    password = config[DOMAIN].get(CONF_PASSWORD)
+
+    state = hass.data[DATA_KEY] = JLRData(config)
+
+    interval = config[DOMAIN].get(CONF_SCAN_INTERVAL)
+
+    connection = jlrpy.Connection(username, password)
+    vehicles = connection.vehicles
+
+    vehicles = []
+    for vehicle in connection.vehicles:
+        __LOGGER.info("Populating vehicle info")
+        vehicle.info = vehicle.get_status()
+        vehicles.append(vehicle)
+
+    def discover_vehicle(vehicle):
+        state.entities[vehicle.vin] = []
+
+        for attr, (component, *_) in RESOURCES.items():
+            hass.helpers.discovery.load_platform(
+                component, DOMAIN, (vehicle.vin, attr), config
+            )
+
+    def update_vehicle(vehicle):
+        """Update information on vehicle."""
+        __LOGGER.info("Pulling info from JLR")
+
+        state.vehicles[vehicle.vin] = vehicle
+        if vehicle.vin not in state.entities:
+            discover_vehicle(vehicle)
+
+        for entity in state.entities[vehicle.vin]:
+            entity.schedule_update_ha_state()
+
+        dispatcher_send(hass, SIGNAL_VEHICLE_SEEN, vehicle)
+
+    def update(now):
+        """Update status from the online service."""
+        __LOGGER.info("Update method in INIT")
+        try:
+            if not connection:
+                __LOGGER.warning("Could not get data from service")
+                return False
+
+            for vehicle in vehicles:
+                update_vehicle(vehicle)
+
+            return True
+        finally:
+            track_point_in_utc_time(hass, update,
+                                    utcnow() + interval)
+
+    __LOGGER.info("Logging into InControl")
+
+    return update(utcnow())
+
+
+class JLRData:
+    """Hold component state."""
+
+    def __init__(self, config):
+        """Initialize the component state."""
+        self.entities = {}
+        self.vehicles = {}
+        self.config = config[DOMAIN]
+        self.names = self.config.get(CONF_NAME)
+
+    def vehicle_name(self, vehicle):
+        """Provide a friendly name for a vehicle."""
+        if vehicle.vin and vehicle.vin.lower() in self.names:
+            return self.names[vehicle.vin.lower()]
+
+        if vehicle.vin:
+            return vehicle.vin
+
+        return ''
+
+
+class JLREntity(Entity):
+    """Base class for all JLR Vehicle entities."""
+
+    def __init__(self, hass, vin, attribute):
+        """Initialize the entity."""
+        self._hass = hass
+        self._vin = vin
+        self._attribute = attribute
+        self._state.entities[self._vin].append(self)
+
+    @staticmethod
+    def _get_vehicle_status(vehicle):
+        dict_only = {}
+        for element in vehicle.get_status().get('vehicleStatus'):
+            dict_only[element.get('key')] = element.get('value')
```

Comment:
Could the key look up return `None`? Then there might be a key clash in the dict.